Simbody/src/SimbodyMatterSubsystemRep.cpp

```diff
@@ -1095,6 +1088,114 @@ int SimbodyMatterSubsystemRep::realizeSubsystemPositionImpl(const State& s) cons
 }
 
 
+//==============================================================================
+//                       REALIZE POSITION KINEMATICS
+//==============================================================================
+void SimbodyMatterSubsystemRep::
+realizePositionKinematics(const State& state) const {
+    const CacheEntryIndex tpcx = topologyCache.treePositionCacheIndex;
+
+    StageVersion posKinVersion;
+    const StageVersion qVersion = state.getQStageVersion();
+    if (isCacheValueRealizedWithExtras(state, tpcx, 1, &qVersion, 
+                                       posKinVersion))
+        return; // already realized
+
+    SimTK_STAGECHECK_GE_ALWAYS(getStage(state), Stage::Instance, 
+        "SimbodyMatterSubsystem::realizePositionKinematics()");
+
+    const SBStateDigest     stateDigest(state, *this, Stage::Time);
+    const SBInstanceVars&   iv = stateDigest.getInstanceVars();
+    SBTreePositionCache&    tpc = stateDigest.updTreePositionCache();
+
+    // realize tree positions (kinematics)
+    // This includes all local cross-mobilizer kinematics (M in F, B in P)
+    // and all global kinematics relative to Ground (G).
+
+    // Any body which is using quaternions should calculate the quaternion
+    // constraint here and put it in the appropriate slot of qErr.
+    // Set generalized coordinates: sweep from base to tips.
+    for (int i=0 ; i<(int)rbNodeLevels.size() ; i++) 
```

Comment:
> ++i?

That would be ideal, but sometimes discretion is the better part of valor. That is very old code and there are many almost-identical doubly-nested loops written identically in that code (the spacing isn't to-spec either). I did notice those and considered hunting them down and cleaning them all up but it seemed like a distraction from the current project. I'm going to leave those for another day.